svctools/man/svc2aut.4
.\"  SVC tools -- the SVC (Systems Validation Centre) tool set
.\"
.\"  Copyright (C) 2000  Stichting Mathematisch Centrum, Amsterdam,
.\"                      The  Netherlands
.\"
.\"  This program is free software; you can redistribute it and/or
.\"  modify it under the terms of the GNU General Public License
.\"  as published by the Free Software Foundation; either version 2
.\"  of the License, or (at your option) any later version.
.\"
.\"  This program is distributed in the hope that it will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"  GNU General Public License for more details.
.\"
.\"  You should have received a copy of the GNU General Public License
.\"  along with this program; if not, write to the Free Software
.\"  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
.\"
.\" $Id: svc2aut.4,v 1.2 2001/01/04 15:55:42 izak Exp $
.TH SVC2AUT 1 15/5/2000
.SH NAME
svc2aut \- file conversion from SVC format to Aldebaran format

.SH SYNOPSIS
.B svc2aut
[-c][-v][-s][-h][-o 
.I 
outfile\c
] 
.I 
infile

.SH DESCRIPTION
.B svc2aut
is a tool for the conversion of files in SVC format (.svc) to files in
Aldebaran format (.aut)

.SH OPTIONS 
.IP \-c
Don't overwrite existing files
.IP \-v
Print version number
.IP \-h
List available options
.IP \-o\ \fIoutfile\fR
Write the output to \fIoutfile\fR. The default is 
.I outfile.aut
.IP \-s
Silent operation: don't print the header information from the input file

.SH FILES
.IP filename.aut
A file in Aldebaran format
.IP filename.svc
A file in SVC format

.SH SEE ALSO

aut2svc(4)

.SH AUTHOR
Izak van Langevelde <izak@cwi.nl>
.LP
Please send all complaints, comments and bug fixes to me. 


svc/ChangeLog
   SVC -- the SVC (Systems Validation Centre) file format library

   Copyright (C) 2000  Stichting Mathematisch Centrum, Amsterdam,
                       The  Netherlands

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public License
   as published by the Free Software Foundation; either version 2
   of the License, or (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

$Id: ChangeLog,v 1.1.1.1 2004/09/07 15:06:33 uid523 Exp $



2000/6/1
   * Version 1.0 
   * Initial release

2000/6/15 (izak@cwi.nl)
   * Version 1.1 
   * Bug fix: frequency of the escape sequence is not updated

2000/9/18 (izak@cwi.nl)
   * Version 1.2
   * Update conform update ATerm library 1.4.4
   * Added configure.in to release

2000/9/25
   * Release 1.0.3
   * Fixed bug in troff manuals causing manuals to start with a blank page
   * Fixed memory allocation bug in compressed streams
   * Cleaned up header files by removing standard includes

2000/9/27
   * Release 1.0.4
   * Fixed bug: buffer for states and labels was allocated too small

2000/10/18
   * Release 1.0.5
   * Fixed bug: the SVCaterm2... functions returned 0 instead of -1 to flag
     an unknown ATerm

2001/1/4
   * Release 1.0.6
   * Format version number incremented to 1.2 beta
   * Lempel Ziff token enlarged to three bytes
   * Gnu General Public License added


svctools/src/svc2aut.c
/*
   SVC tools -- the SVC (Systems Validation Centre) tool set

   Copyright (C) 2000  Stichting Mathematisch Centrum, Amsterdam,
                       The  Netherlands

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public License
   as published by the Free Software Foundation; either version 2
   of the License, or (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

 $Id: svc2aut.c,v 1.1.1.1 2004/09/07 15:06:33 uid523 Exp $ */

#include "svc2aut.h"



int main(int argc, char *argv[]) {
   FILE *fpOut;
   int traceLevel;
   ATerm bottom;


   ATinit(argc, argv, &bottom);

   switch(parseArgs(argc, argv,&fpOut,&traceLevel)) {
      case ERR_ARGS:
         doHelp(argv[0]);
         exit(EXIT_ERR_ARGS);
         break;
      case ERR_FILE:
         exit(EXIT_ERR_FILE);
         break;
      case CMD_HELP:
         doHelp(argv[0]);
         exit(EXIT_OK);
         break;
      case CMD_VERSION:
         doVersion(argv[0]);
         exit(EXIT_OK);
      case CMD_CONVERT:
         exit(doConvert(fpOut, traceLevel));
         break;
      default:
         exit(EXIT_OK);
   }


} /* main */



int parseArgs(int argc, char *argv[], FILE **fpOut, int *traceLevel){
   int c, ret, cautious;
   extern int optind;
   char *inFilename=NULL, *outFilename=NULL;
   SVCbool indexed, allocatedFileName = SVCfalse;


   *traceLevel=1;
   cautious=0;
   while ((c = getopt(argc, argv, "cho:sv")) != EOF) {
      switch(c){
         case 'c': cautious=1;
                   break;
         case 'h': return CMD_HELP;
         case 'o': outFilename=optarg;
                   break;
         case 's': *traceLevel=0;
                   break;
         case 'v': return CMD_VERSION;
         case '?': return ERR_ARGS;
      }
   }

   if (optind == argc-1) {

      /* Open the filename given as argument */

      inFilename = (char*)malloc(sizeof(char)*(strlen(argv[optind])+strlen(INFILE_EXT)+2));
      strcpy(inFilename, argv[optind]);

      if (SVCopen(&inFile, inFilename, SVCread, &indexed) ==0) {
         ret= CMD_CONVERT;
      } else {

         /* Open the filename given as argument with extension */

         if (SVCerrno==EACCESS){

            sprintf(inFilename,"%s%s", argv[optind],INFILE_EXT);

            if (SVCopen(&inFile, inFilename, SVCread, &indexed) >= 0) {
               ret= CMD_CONVERT;
            } else {
               fprintf(stderr, "%s: %s\n", argv[optind], SVCerror(SVCerrno));
               ret= ERR_FILE;
            }
         } else{
            fprintf(stderr, "%s: %s\n", argv[optind], SVCerror(SVCerrno));
            ret= ERR_FILE;
         }

      }

   } else {

      /* No filename is given as argument: this is an error */

      ret=ERR_ARGS;

   }

   if (ret==CMD_CONVERT) {

      if (outFilename==NULL){

         /* Remove the extension from the input filename */

         if(strlen(inFilename)>strlen(INFILE_EXT) &&
            strcmp(inFilename+strlen(inFilename)-strlen(INFILE_EXT),INFILE_EXT)==0) {
            inFilename[strlen(inFilename)-strlen(INFILE_EXT)]='\0';
         } 

         /* Compose output filename */

         outFilename=(char*) malloc(sizeof(char)*(strlen(inFilename)+strlen(OUTFILE_EXT)+1));
         allocatedFileName = SVCtrue;
         sprintf(outFilename,"%s%s",inFilename,OUTFILE_EXT);

      }

      if (cautious && access(outFilename,F_OK)==0) {
         fprintf(stderr, "%s: file already exists\n", outFilename);
         ret=ERR_FILE;
      } else {
         if ((*fpOut=fopen(outFilename,"w"))!=NULL) {
            ret= CMD_CONVERT;
         } else {
            fprintf(stderr, "%s: %s\n", outFilename,strerror(errno));
            ret= ERR_FILE;
         }
      }

   }

   free(inFilename);
   if (allocatedFileName) free(outFilename);

   return ret;

} /* parseArgs */



void doHelp(char *cmd) {

   fprintf(stdout, "Usage: %s [-c][-v][-h][-s][-o outfile] infile\n", cmd);
   fprintf(stdout, "\n");
   fprintf(stdout, "Flags:\n");
   fprintf(stdout, "-c  Cautious mode: don't overwrite existing files\n");
   fprintf(stdout, "-v  Print version number\n");
   fprintf(stdout, "-h  Print this help info\n");
   fprintf(stdout, "-s  Silent: no logging is printed\n");
   fprintf(stdout, "-o  Output to `outfile'\n");


} /* doHelp */



void doVersion(char *progName) {

   fprintf(stdout, "%s %s (%s)\n", progName, VERSION, DATE);

} /* doVersion */



int doConvert(FILE *fpOut, int traceLevel) {
   SVCstateIndex fromState, toState;
   SVClabelIndex label;
   SVCparameterIndex parameter;

   if (traceLevel>0){

      /* Get file header info */

      fprintf(stderr, "Svc version %s\n", SVCgetVersion(&inFile));
      fprintf(stderr, " filename      %s\n", SVCgetFilename(&inFile));
      fprintf(stderr, " date          %s\n", SVCgetDate(&inFile));
      fprintf(stderr, " version       %s\n", SVCgetVersion(&inFile));
      fprintf(stderr, " type          %s\n", SVCgetType(&inFile));
      fprintf(stderr, " creator       %s\n",  SVCgetCreator(&inFile));
      fprintf(stderr, " states        %ld\n", SVCnumStates(&inFile));
      fprintf(stderr, " transitions   %ld\n", SVCnumTransitions(&inFile));
      fprintf(stderr, " labels        %ld\n", SVCnumLabels(&inFile));
      fprintf(stderr, " parameters    %ld\n", SVCnumParameters(&inFile));
    ATfprintf(stderr, " initial state %t\n", SVCstate2ATerm(&inFile,SVCgetInitialState(&inFile)));
      fprintf(stderr, " comments      %s\n", SVCgetComments(&inFile));
   }

   fprintf(fpOut, "des(%ld,%ld,%ld)\n", SVCgetInitialState(&inFile),
                                      SVCnumTransitions(&inFile),
                                      SVCnumStates(&inFile));

   while (SVCgetNextTransition(&inFile, &fromState, &label, &toState, &parameter)) {
      ATfprintf(fpOut, "(%d, %t, %d)\n", fromState, SVClabel2ATerm(&inFile,label), toState);
   }

   if (SVCclose(&inFile)<0){
      if (traceLevel>0) {
         fprintf(stderr, "File trailer corrupt...\n");
      }
      return -1;
   }

   fclose(fpOut);
   return EXIT_OK;

} /* doConvert */
